File: scripts/skillopt-eval/contracts/review.ts
import { z } from "zod";
import {
  CONTRACT_SCHEMA_VERSION,
  ContractIntegrityError,
  JsonValueSchema,
  NonEmptyStringSchema,
  Sha256Schema,
  TimestampSchema,
  boundedContractSchema,
  contractHash,
} from "./common";
import { SkillSchema } from "./episode";

// implements REQ-skillopt-codex-optimization
export const ProposalSchema = boundedContractSchema(
  z
    .object({
      schemaVersion: z.literal(CONTRACT_SCHEMA_VERSION),
      artifactType: z.literal("proposal"),
      proposalId: z.uuid(),
      runId: z.uuid(),
      runLockHash: Sha256Schema,
      skill: SkillSchema,
      candidateBodyHash: Sha256Schema,
      baselineFrontmatterHash: Sha256Schema,
      candidateFrontmatterHash: Sha256Schema,
      baselineResourcesHash: Sha256Schema,
      candidateResourcesHash: Sha256Schema,
      reportHash: Sha256Schema,
      createdAt: TimestampSchema,
      status: z.enum(["eligible", "accepted", "rejected"]),
    })
    .strict()
    .superRefine((proposal, context) => {
      if (
        proposal.baselineFrontmatterHash !== proposal.candidateFrontmatterHash
      ) {
        context.addIssue({
          code: "custom",
          message: "frontmatter hash changed",
        });
      }
      if (proposal.baselineResourcesHash !== proposal.candidateResourcesHash) {
        context.addIssue({ code: "custom", message: "resources hash changed" });
      }
    }),
);

// implements REQ-skillopt-codex-optimization
export const ApprovalSchema = boundedContractSchema(
  z
    .object({
      schemaVersion: z.literal(CONTRACT_SCHEMA_VERSION),
      artifactType: z.literal("approval"),
      approvalId: z.uuid(),
      proposalId: z.uuid(),
      proposalHash: Sha256Schema,
      runId: z.uuid(),
      runLockHash: Sha256Schema,
      reportHash: Sha256Schema,
      candidateBodyHash: Sha256Schema,
      reviewer: NonEmptyStringSchema,
      decision: z.literal("approved"),
      decidedAt: TimestampSchema,
    })
    .strict(),
);

// implements REQ-skillopt-codex-optimization
export type Proposal = Readonly<z.infer<typeof ProposalSchema>>;
// implements REQ-skillopt-codex-optimization
export type Approval = Readonly<z.infer<typeof ApprovalSchema>>;

// implements REQ-skillopt-codex-optimization
export function assertApprovalMatchesProposal(
  proposal: Proposal,
  approval: Approval,
): void {
  if (proposal.status !== "eligible" && proposal.status !== "accepted") {
    throw new ContractIntegrityError(
      "proposal is not approval-eligible",
      "proposal.status",
    );
  }
  const proposalHash = contractHash(JsonValueSchema.parse(proposal));
  if (
    approval.proposalId !== proposal.proposalId ||
    approval.runId !== proposal.runId ||
    approval.runLockHash !== proposal.runLockHash ||
    approval.reportHash !== proposal.reportHash ||
    approval.candidateBodyHash !== proposal.candidateBodyHash ||
    approval.proposalHash !== proposalHash
  ) {
    throw new ContractIntegrityError(
      "approval does not match proposal",
      "approval",
    );
  }
}

File: tools/skillopt/kibi_skillopt/review.py
from __future__ import annotations

from typing import Annotated, Literal
from uuid import UUID

from pydantic import AwareDatetime, Field, model_validator
from typing_extensions import Self

from .common import (
    ContractModel,
    ContractValidationError,
    NonEmptyString,
    Sha256,
    contract_hash,
    parse_json_value,
)
from .episode import Skill


class Proposal(ContractModel):
    schema_version: Annotated[Literal["1.0.0"], Field(alias="schemaVersion")]
    artifact_type: Annotated[Literal["proposal"], Field(alias="artifactType")]
    proposal_id: Annotated[UUID, Field(alias="proposalId")]
    run_id: Annotated[UUID, Field(alias="runId")]
    run_lock_hash: Annotated[Sha256, Field(alias="runLockHash")]
    skill: Skill
    candidate_body_hash: Annotated[Sha256, Field(alias="candidateBodyHash")]
    baseline_frontmatter_hash: Annotated[Sha256, Field(alias="baselineFrontmatterHash")]
    candidate_frontmatter_hash: Annotated[Sha256, Field(alias="candidateFrontmatterHash")]
    baseline_resources_hash: Annotated[Sha256, Field(alias="baselineResourcesHash")]
    candidate_resources_hash: Annotated[Sha256, Field(alias="candidateResourcesHash")]
    report_hash: Annotated[Sha256, Field(alias="reportHash")]
    created_at: Annotated[AwareDatetime, Field(alias="createdAt")]
    status: Literal["eligible", "accepted", "rejected"]

    @model_validator(mode="after")
    def verify_immutable_surfaces(self) -> Self:
        if self.baseline_frontmatter_hash != self.candidate_frontmatter_hash:
            raise ContractValidationError("frontmatter hash changed")
        if self.baseline_resources_hash != self.candidate_resources_hash:
            raise ContractValidationError("resources hash changed")
        return self


class Approval(ContractModel):
    schema_version: Annotated[Literal["1.0.0"], Field(alias="schemaVersion")]
    artifact_type: Annotated[Literal["approval"], Field(alias="artifactType")]
    approval_id: Annotated[UUID, Field(alias="approvalId")]
    proposal_id: Annotated[UUID, Field(alias="proposalId")]
    proposal_hash: Annotated[Sha256, Field(alias="proposalHash")]
    run_id: Annotated[UUID, Field(alias="runId")]
    run_lock_hash: Annotated[Sha256, Field(alias="runLockHash")]
    report_hash: Annotated[Sha256, Field(alias="reportHash")]
    candidate_body_hash: Annotated[Sha256, Field(alias="candidateBodyHash")]
    reviewer: NonEmptyString
    decision: Literal["approved"]
    decided_at: Annotated[AwareDatetime, Field(alias="decidedAt")]


def assert_approval_matches_proposal(proposal: Proposal, approval: Approval) -> None:
    if proposal.status not in ("eligible", "accepted"):
        raise ContractValidationError("proposal is not approval-eligible")
    matches = (
        approval.proposal_id == proposal.proposal_id
        and approval.run_id == proposal.run_id
        and approval.run_lock_hash == proposal.run_lock_hash
        and approval.report_hash == proposal.report_hash
        and approval.candidate_body_hash == proposal.candidate_body_hash
        and approval.proposal_hash
        == contract_hash(parse_json_value(proposal.model_dump_json(by_alias=True)))
    )
    if not matches:
        raise ContractValidationError("approval does not match proposal")
